replace2.js

```diff
@@ -1,869 +1,146 @@
-  const current_date = new Date()
-  
-  // first... build grids...
-  contents = contents.replace(grid_replace_regex, parseGrid);
-  // then we must unroll the loops...
-  contents = unrollLoops(contents);
-  // #issue 378, note: getMonth 0=Jan,  need to add 1
-  // months
-
-  contents = contents
-    .replace(/#currentMonthStr/g, moduleParams.i18n.months[new Date().getMonth()])
-    .replace(/#currentMonth/g, current_date.getMonth() + 1)
-    .replace(/#currentYear/g, current_date.getFullYear())
-    // issue #405 need #today and today+/- n days...
-    .replace(/#today(\s*[+\-]\s*\d+)?/g, (match, offset) => {
-      // if no (+/- offset) we want today...
-      if (!offset || offset.trim().length == 0) {
-        return current_date.toQuestFormat()
-      }
-
-      // otherwise +/- the offset in number of days...
-      offset = parseInt(offset.replace(/\s/g, ""));
-      let offset_date = new Date()
-      offset_date.setDate(offset_date.getDate() + offset)
-      return offset_date.toQuestFormat()
-    })
-    // questionnarie 
-    // hey, lets de-lint the contents.. convert (^|\n{2,}Q1. to [Q1]
-    // note:  the first question wont have the \n\n so we need to look at start of string(^)
-    .replace(/\/\*.*\*\//g, "")
-    .replace(/\/\/.*/g, "");
-
-  // contents = contents.replace(/\[DISPLAY IF .*\]/gms, "");
-  let nameRegex = new RegExp(/{"name":"(\w*)"}/);
-  if (nameRegex.test(contents)) {
-    contents = contents.replace(/{"name":"(\w*)"}/, fQuestName);
-    function fQuestName(group, name) {
-      questName = name;
-      moduleParams.questName = name;
-      return "";
-    }
-  } else {
-    questName = "Questionnaire";
-    moduleParams.questName = questName;
-  }
-
-  // first let's deal with breaking up questions..
-  // a question starts with the [ID1] regex pattern
-  // and end with the next pattern or the end of string...
-
-  // start with a '['
-  // then the first character must be a capital letter
-  // followed by zero or more capital letters/digits or an _
-  // note: we want this possessive (NOT greedy) so add a ?
-  //       otherwise it would match the first and last square bracket
-
-  let regEx = new RegExp(
-    "\\[([A-Z_][A-Z0-9_#]*[\\?\\!]?)(?:\\|([^,\\|\\]]+)\\|?)?(,.*?)?\\](.*?)(?=$|\\[[_A-Z]|<form)",
-    "g"
-  );
-
-  // TODO: General -> <br> tag replacement -> these read "Empty Group" in Chrome with screen reader, which is not ideal.
-  // because firefox cannot handle the "s" tag, encode all newlines
-  // as a unit seperator ASCII code 1f (decimal: 31)
-  contents = contents.replace(/(?:\r\n|\r|\n)/g, "\u001f");
-  contents = contents.replace(regEx, function (
-    page,
-    questID,
-    questOpts,
-    questArgs,
-    questText
-  ) {
+  // if the object has a 'text' field, the contents have been prefetched and passed in. Else, fetch the survey contents.
+  const contents = moduleParams.renderObj?.text || await fetch(moduleParams.renderObj?.url).then(response => response.text());
```

Comment:
do we need to null check URL?

Reply:
Thanks for pointing this out. I like the idea of checking it for others using the system -- in case that param is forgotten at integration time. I added a handler directly below this line for null/empty contents.